File: gke_customization/gke_catalog/report/consumables_issue_to_employee_report/consumables_issue_to_employee_report.py
# Copyright (c) 2026, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    """Columns for summary view"""
    return [
        {
            "fieldname": "worker",
            "label": _("Worker"),
            "fieldtype": "Data",
            "width": 350
        },
        {
            "fieldname": "total_amount",
            "label": _("Total Amount"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "view_details",
            "label": _("Details"),
            "fieldtype": "Data",
            "width": 150
        }
    ]

def get_data(filters):
    """Get summary data grouped by employee"""
    
    conditions = get_conditions(filters)
    
    query = f"""
        SELECT 
            CONCAT(emp.employee_name, ' (', se.custom_employee, ')') as worker,
            se.custom_employee as worker_code,
            SUM(sed.qty * COALESCE(pri.rate, 0)) as total_amount,
            COUNT(DISTINCT se.name) as total_transactions
        FROM 
            `tabStock Entry` se
        LEFT JOIN `tabStock Entry Detail` sed ON sed.parent = se.name
        LEFT JOIN `tabEmployee` emp ON emp.name = se.custom_employee
        LEFT JOIN `tabBatch` batch ON batch.name = sed.batch_no
        LEFT JOIN `tabPurchase Receipt Item` pri ON pri.parent = batch.reference_name 
            AND pri.item_code = sed.item_code 
            AND batch.reference_doctype = 'Purchase Receipt'
        WHERE 
            se.stock_entry_type = 'Consumables Issue to  Employee'
            AND se.docstatus = 1
            AND se.custom_employee IS NOT NULL
            {conditions}
        GROUP BY se.custom_employee, emp.employee_name
        ORDER BY total_amount DESC
    """
    
    data = frappe.db.sql(query, filters, as_dict=1)
    
    # Add view details button HTML
    for row in data:
        row['view_details'] = f'Details'
    
    return data

def get_conditions(filters):
    """Build WHERE conditions from filters"""
    conditions = []
    
    if filters.get("from_date"):
        conditions.append(f"""se.posting_date >= "{filters['from_date']}" """)
    
    if filters.get("to_date"):
        conditions.append(f"""se.posting_date <= "{filters['to_date']}" """)
    
    if filters.get("department"):
        departments = ', '.join([f'"{dept}"' for dept in filters.get("department")])
        conditions.append(f"emp.department IN ({departments})")
    
    return " AND " + " AND ".join(conditions) if conditions else ""

@frappe.whitelist()
def get_employee_details(employee_code, from_date=None, to_date=None):
    """API method to fetch employee details for popup"""
    
    conditions = []
    if from_date:
        conditions.append(f"""se.posting_date >= "{from_date}" """)
    if to_date:
        conditions.append(f"""se.posting_date <= "{to_date}" """)
    
    conditions.append(f"""se.custom_employee = "{employee_code}" """)
    
    conditions_sql = " AND " + " AND ".join(conditions)
    
    query = f"""
        SELECT 
            se.name as stock_entry,
            se.posting_date as issue_date,
            emp.department as dept,
            mgr.employee_name as manager,
            emp.employee_name as worker,
            se.custom_employee as worker_code,
            sed.item_code,
            item.item_name,
            item.item_group,
            item.custom_old_item_category as category,
            item.custom_old_item_sub_category as sub_category,
            sed.qty,
            pri.rate,
            sed.stock_uom as unit,
            (sed.qty * COALESCE(pri.rate, 0)) as total,
            batch.name as batch_no,
            batch.reference_name as purchase_receipt,
            pr.posting_date as pr_date
        FROM 
            `tabStock Entry` se
        LEFT JOIN `tabStock Entry Detail` sed ON sed.parent = se.name
        LEFT JOIN `tabEmployee` emp ON emp.name = se.custom_employee
        LEFT JOIN `tabEmployee` mgr ON mgr.name = emp.reports_to
        LEFT JOIN `tabItem` item ON item.name = sed.item_code
        LEFT JOIN `tabBatch` batch ON batch.name = sed.batch_no
        LEFT JOIN `tabPurchase Receipt` pr ON pr.name = batch.reference_name 
            AND batch.reference_doctype = 'Purchase Receipt'
        LEFT JOIN `tabPurchase Receipt Item` pri ON pri.parent = batch.reference_name 
            AND pri.item_code = sed.item_code
        WHERE 
            se.stock_entry_type = 'Consumables Issue to  Employee'
            AND se.docstatus = 1
            {conditions_sql}
        ORDER BY se.posting_date DESC, se.name
    """
    
    data = frappe.db.sql(query, as_dict=1)
    
    # Add serial numbers
    for idx, row in enumerate(data, 1):
        row['serial_no'] = idx
    
    return data

File: gke_customization/gke_catalog/report/consumables_issue_to_employee_report/consumables_issue_to_employee_report.js
// Copyright (c) 2026, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Consumables Issue to Employee Report"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 0
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 0
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "MultiSelectList",
			options: "Department",
			reqd: 0,
			get_data: function(txt) {
				return frappe.db.get_link_options("Department", txt);
			}
		}
	],

	onload: function(report) {
		const clearFilterButton = report.page.add_inner_button(__("Clear Filter"), function() {
			report.filters.forEach(function(filter) {
				let field = report.get_filter(filter.fieldname);

				if (field.df.fieldtype === "MultiSelectList") {
					field.set_value([]);
				} else if (field.df.default) {
					field.set_value(field.df.default);
				} else {
					field.set_value("");
				}
			});
		});
	},

	formatter: function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Make view_details clickable
		if (column.fieldname == "view_details" && data && data.worker_code) {
			return `<a onclick="showEmployeeDetails('${data.worker_code}', '${data.worker}')" style="cursor:pointer; color: #2490ef;">${value}</a>`;
		}
		
		return value;
	}
};

// Global function to show employee details in dialog
window.showEmployeeDetails = function(employee_code, worker_name) {
	let filters = frappe.query_report.get_filter_values();
	
	frappe.call({
		method: 'gke_customization.gke_catalog.report.consumables_issue_to_employee_report.consumables_issue_to_employee_report.get_employee_details',
		args: {
			employee_code: employee_code,
			from_date: filters.from_date,
			to_date: filters.to_date
		},
		callback: function(r) {
			if (r.message) {
				let data = r.message;
				
				// Create HTML table
				let html = `
					<div style="max-height: 500px; overflow-y: auto;">
						<table class="table table-bordered table-hover">
							<thead style="position: sticky; top: 0; background: white;">
								<tr>
									<th>Sr No</th>
									<th>Issue Date</th>
									<th>Dept</th>
									<th>Manager</th>
									<th>Item Code</th>
									<th>Item Group</th>
									<th>Category</th>
									<th>Sub Category</th>
									<th>Qty</th>
									<th>Rate</th>
									<th>Unit</th>
									<th>Total</th>
								</tr>
							</thead>
							<tbody>
				`;
				
				data.forEach(row => {
					html += `
						<tr>
							<td>${row.serial_no}</td>
							<td>${frappe.datetime.str_to_user(row.issue_date)}</td>
							<td>${row.dept || ''}</td>
							<td>${row.manager || ''}</td>
							<td>${row.item_code || ''}</td>
							<td>${row.item_group || ''}</td>
							<td>${row.category || ''}</td>
							<td>${row.sub_category || ''}</td>
							<td>${row.qty}</td>
							<td>${format_currency(row.rate)}</td>
							<td>${row.unit}</td>
							<td>${format_currency(row.total)}</td>
						</tr>
					`;
				});
				
				html += `
							</tbody>
						</table>
					</div>
				`;
				
				// Show dialog
				let d = new frappe.ui.Dialog({
					title: __('Issue Details - {0}', [worker_name]),
					fields: [
						{
							fieldtype: 'HTML',
							fieldname: 'details_html'
						}
					],
					size: 'extra-large'
				});
				
				d.fields_dict.details_html.$wrapper.html(html);
				d.show();
			}
		}
	});
};
